redis/src/lib.rs

```diff
@@ -0,0 +1,100 @@
+//! Redis support for the `bb8` connection pool.
+#![deny(missing_docs, missing_debug_implementations)]
+
+pub extern crate bb8;
+pub extern crate redis;
+
+extern crate futures;
+extern crate tokio_core;
+
+use futures::{Future, IntoFuture};
+use tokio_core::reactor::Handle;
+
+use redis::async::Connection;
+use redis::{Client, RedisError};
+
+use std::io;
+use std::option::Option;
+
+type Result<T> = std::result::Result<T, RedisError>;
+
+/// `RedisPool` is a convenience wrapper around `bb8::Pool` that hides the fact that
+/// `RedisConnectionManager` uses an `Option<Connection>` to smooth over the API incompatibility.
+#[derive(Debug)]
+pub struct RedisPool {
+    // The `bb8::Builder` that manages the pool.
+    pool: bb8::Pool<RedisConnectionManager>,
+}
+
+impl RedisPool {
+    /// Constructs a new `RedisPool`, see the `bb8::Builder` documentation for description of
+    /// parameters.
+    pub fn new(pool: bb8::Pool<RedisConnectionManager>) -> RedisPool {
+        RedisPool { pool }
+    }
+
+    /// Run the function with a connection provided by the pool.
+    pub fn run<'a, T, E, U, F>(&self, f: F) -> Box<Future<Item = T, Error = E> + Send + 'a>
+    where
+        F: FnOnce(Connection) -> U + Send + 'a,
+        U: IntoFuture<Item = (Connection, T), Error = E> + 'a,
+        U::Future: Send + 'a,
+        E: From<RedisError> + Send + 'a,
+        T: Send + 'a,
+    {
+        let f = move |conn: Option<Connection>| {
+            let conn = conn.unwrap();
```

Comment:
Wouldn't it make more sense to throw an error here than unwrap (and potentially panic)?

Reply:
I may have misunderstood the invariants provided by the bb8 library. My understanding was that this method would never get called with a connection after an error. Unwrapping both makes the code simpler and effectively serves as an assert. This will only ever be `None` if there was already a failed call to `run` or `is_valid`. I can change it to be an error if my understanding is incorrect or you'd rather we return an error regardless.

Reply:
In order to get here the connection must be either a) freshly generated from `add_connection` or b) returned from a previous `Pool::run` call. In the first case, `RedisConnectionManager::connect` always returns `Some`, and in the second case `RedisConnectionManager::has_broken` would have rejected any connection that is `None` before we put it back in the pool, so I believe you are correct here.